lib/auth/userloginstate/generator.go

```diff
@@ -0,0 +1,175 @@
+	"github.com/gravitational/teleport/api/types/trait"
+	"github.com/gravitational/teleport/api/types/userloginstate"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/services/local"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// GeneratorConfig is the configuration for the user login state generator.
+type GeneratorConfig struct {
+	// AccessLists is a service for retrieving access lists from the backend.
+	AccessLists services.AccessListsGetter
+
+	// Roles is a service for retrieving roles from the backend.
+	Roles services.RoleGetter
+
+	// Clock is the clock to use for the generator.
+	Clock clockwork.Clock
+}
+
+func (g *GeneratorConfig) CheckAndSetDefaults() error {
+	if g.AccessLists == nil {
+		return trace.BadParameter("missing access lists")
+	}
+
+	if g.Roles == nil {
+		return trace.BadParameter("missing roles")
+	}
+
+	if g.Clock == nil {
+		g.Clock = clockwork.NewRealClock()
+	}
+
+	return nil
+}
+
+// Generator will generate a user login state from a user.
+type Generator struct {
+	accessLists services.AccessListsGetter
+	roles       services.RoleGetter
+	clock       clockwork.Clock
+}
+
+// NewGenerator creates a new user login state generator.
+func NewGenerator(config GeneratorConfig) (*Generator, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Generator{
+		accessLists: config.AccessLists,
+		roles:       config.Roles,
+		clock:       config.Clock,
+	}, nil
+}
+
+// Generate will generate the user login state for the given user.
+func (g *Generator) Generate(ctx context.Context, user types.User) (*userloginstate.UserLoginState, error) {
+	// Create a new empty user login state.
+	uls, err := userloginstate.New(
+		header.Metadata{
+			Name: user.GetName(),
+		}, userloginstate.Spec{
+			Roles:  []string{},
+			Traits: trait.Traits{},
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Generate the user login state.
+	if err := g.addAccessListsToState(ctx, user, uls); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Clean up the user login state after generating it.
+	if err := g.postProcess(ctx, uls); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return uls, nil
+}
+
+// addAccessListsToState will added the user's applicable access lists to the user login state.
+func (g *Generator) addAccessListsToState(ctx context.Context, user types.User, state *userloginstate.UserLoginState) error {
+	accessLists, err := g.accessLists.GetAccessLists(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Create an identity for testing membership to access lists.
+	identity := tlsca.Identity{
+		Username: user.GetName(),
+		Groups:   user.GetRoles(),
+		Traits:   user.GetTraits(),
+	}
+
+	for _, accessList := range accessLists {
+		if err := services.IsMember(identity, g.clock, accessList); err != nil {
+			continue
```

Comment:
The order in which we process elements of `accessLists` can affect the end result of the function here. Specifically, we can have a situation where access list `foo` will effectively grant access for `bar`, but only if it is processed first. 

I think we can fix this fairly easily: if user was found to be a member of any of the lists, retry the whole loop again (but only for rejected items).

This can blow up spectacularly with large or malicious access lists, but we can try to optimize it as needed.

Reply:
So I don't think this is actually the case here -- access list membership requirements is dependent on the user's fixed definition, not the emerging definition.

Reply:
Hmm, I'll dig into this and check my theory.

Reply:
Ok, I see I was wrong with this. This does simplify things, although I wonder if we will ever come to change this.